File: 프로그래머스/lv2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.cpp
#include <string>
#include <vector>
#include <deque>
#include <numeric>

using namespace std;

int solution(vector<int> queue1, vector<int> queue2) {
    long answer = 0;
    deque<int> dequeue1(queue1.begin(), queue1.end());
    deque<int> dequeue2(queue2.begin(), queue2.end());
    long sum1 = accumulate(dequeue1.begin(), dequeue1.end(), 0);
    long sum2 = accumulate(dequeue2.begin(), dequeue2.end(), 0);
    long sum = sum1 + sum2;
    long diff = sum1 - sum2;
    
    if (sum % 2) {
        return -1;
    }
    
    long initial_sum1 = sum1;
    long initial_sum2 = sum2;
    while (diff) {
        if (diff > 0) {
            int temp = dequeue1[0];
            dequeue1.pop_front();
            dequeue2.push_back(temp);
            sum1 -= temp;
            sum2 += temp;
        } else {
            int temp = dequeue2[0];
            dequeue2.pop_front();
            dequeue1.push_back(temp);
            sum1 += temp;
            sum2 -= temp;
        }
        answer++;
        diff = sum1 - sum2;
        if (answer >= queue1.size() * 4) {
            return -1;
        }
    }
    
    return answer;
}

File: 프로그래머스/lv2/118667. 두 큐 합 같게 만들기/두 큐 합 같게 만들기.java
import java.util.*;
import java.io.*;

class Solution {
    public int solution(int[] queue1, int[] queue2) {
        int answer = 0;
        int sum1 = 0;
        int sum2 = 0;
        
        Queue<Integer> q1 = new LinkedList<>();
        Queue<Integer> q2 = new LinkedList<>();
        
        for (int i = 0; i < queue1.length; i++) {
            q1.add(queue1[i]);
            q2.add(queue2[i]);
            sum1 += queue1[i];
            sum2 += queue2[i];
        }
        
        if ((sum1 + sum2) % 2 == 1)
            return -1;
        
        int diff = sum1 - sum2;
        while (diff != 0 && (answer < queue1.length * 4)) {
            if (diff >  0) {
                sum1 -= q1.peek();
                sum2 += q1.peek();
                q2.add(q1.peek());
                q1.poll();
            } else {
                sum1 += q2.peek();
                sum2 -= q2.peek();
                q1.add(q2.peek());
                q2.poll();
            }
            diff = sum1 - sum2;
            answer++;
        }
        if (diff != 0)
            return -1;
        return answer;
    }
}